File: proxy-server.js
const http = require('http');
const https = require('https');
const { URL } = require('url');

const PORT = process.env.PORT || 3000;
const HOST = process.env.HOST || '0.0.0.0';

function sendJson(res, statusCode, payload) {
  const body = JSON.stringify(payload);
  res.writeHead(statusCode, {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Cache-Control': 'no-store',
    'Content-Length': Buffer.byteLength(body),
  });
  res.end(body);
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', (chunk) => {
      body += chunk;
    });
    req.on('end', () => resolve(body));
    req.on('error', reject);
  });
}

function requestUpstream(url, options = {}) {
  return new Promise((resolve, reject) => {
    const parsedUrl = new URL(url);
    const transport = parsedUrl.protocol === 'https:' ? https : http;
    const req = transport.request(parsedUrl, {
      method: options.method || 'GET',
      headers: options.headers || {},
    }, (res) => {
      let body = '';
      res.setEncoding('utf8');
      res.on('data', (chunk) => {
        body += chunk;
      });
      res.on('end', () => {
        resolve({ statusCode: res.statusCode || 0, body });
      });
    });

    req.on('error', reject);
    if (options.body) {
      req.write(options.body);
    }
    req.end();
  });
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url, `http://${req.headers.host || 'localhost'}`);

  if (req.method === 'OPTIONS') {
    res.writeHead(204, {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type',
    });
    res.end();
    return;
  }

  if (url.pathname === '/proxy/nominatim') {
    const query = new URL('https://nominatim.openstreetmap.org/search');
    query.searchParams.set('format', 'jsonv2');
    query.searchParams.set('limit', url.searchParams.get('limit') || '8');
    query.searchParams.set('addressdetails', url.searchParams.get('addressdetails') || '1');
    query.searchParams.set('q', url.searchParams.get('q') || '');

    try {
      const upstreamResponse = await requestUpstream(query.toString(), {
        headers: {
          'Accept-Language': 'en',
          'User-Agent': 'walker-streets-proxy-test',
        },
      });
      if (upstreamResponse.statusCode >= 400) {
        throw new Error(`Nominatim returned ${upstreamResponse.statusCode}`);
      }

      let data;
      try {
        data = JSON.parse(upstreamResponse.body);
      } catch (parseError) {
        throw new Error(`Unable to parse Nominatim response: ${parseError.message}`);
      }

      sendJson(res, 200, data);
    } catch (error) {
      sendJson(res, 500, { error: 'Unable to fetch place suggestions', details: error.message });
    }
    return;
  }

  if (url.pathname === '/proxy/overpass') {
    try {
      const bodyText = await readBody(req);
      const query = (bodyText || '').trim();
      const upstreamUrl = new URL('https://overpass-api.de/api/interpreter');
      if (query) {
        upstreamUrl.searchParams.set('data', query);
      }

      const upstreamResponse = await requestUpstream(upstreamUrl.toString(), {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'text/plain; charset=utf-8',
          'User-Agent': 'walker-streets-proxy/1.0 (+https://streetwalker.onrender.com)',
        },
        body: query,
      });

      if (upstreamResponse.statusCode >= 400) {
        throw new Error(`Overpass returned ${upstreamResponse.statusCode}`);
      }

      const text = upstreamResponse.body;
      if (!text) {
        throw new Error('Upstream returned an empty response');
      }

      let parsed;
      try {
        parsed = JSON.parse(text);
      } catch (parseError) {
        if (text.includes('<!DOCTYPE') || text.includes('<html')) {
          throw new Error(`Overpass returned HTML instead of JSON: ${text.slice(0, 160)}`);
        }
        throw parseError;
      }

      sendJson(res, 200, parsed);
    } catch (error) {
      sendJson(res, 500, { error: 'Unable to fetch street data', details: error.message });
    }
    return;
  }

  if (url.pathname === '/health') {
    sendJson(res, 200, { ok: true, service: 'walker-streets-proxy' });
    return;
  }

  sendJson(res, 404, { error: 'Not found' });
});

server.listen(PORT, HOST, () => {
  console.log(`Proxy server is listening on http://${HOST}:${PORT}`);
});

File: proxy-server.py
import json
import os
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import urlparse, parse_qs
from urllib.request import Request, urlopen

PORT = int(os.environ.get('PORT', '3000'))
HOST = os.environ.get('HOST', '0.0.0.0')


class ProxyHandler(BaseHTTPRequestHandler):
    def _send_json(self, status, payload):
        body = json.dumps(payload).encode('utf-8')
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(body)))
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == '/health':
            self._send_json(200, {'ok': True, 'service': 'walker-streets-proxy'})
            return

        if parsed.path == '/proxy/nominatim':
            params = parse_qs(parsed.query)
            q = params.get('q', [''])[0]
            limit = params.get('limit', ['8'])[0]
            addressdetails = params.get('addressdetails', ['1'])[0]
            upstream_url = 'https://nominatim.openstreetmap.org/search?format=jsonv2&limit=' + limit + '&addressdetails=' + addressdetails + '&q=' + q
            req = Request(upstream_url, headers={'Accept-Language': 'en', 'User-Agent': 'walker-streets-proxy-test'})
            try:
                with urlopen(req, timeout=30) as resp:
                    data = json.load(resp)
                self._send_json(200, data)
            except Exception as exc:
                self._send_json(500, {'error': 'Unable to fetch place suggestions', 'details': str(exc)})
            return

        self._send_json(404, {'error': 'Not found'})

    def do_POST(self):
        parsed = urlparse(self.path)
        if parsed.path != '/proxy/overpass':
            self._send_json(404, {'error': 'Not found'})
            return

        length = int(self.headers.get('Content-Length', '0'))
        body = self.rfile.read(length).decode('utf-8') if length else ''
        query = body.strip()
        upstream_url = 'https://overpass-api.de/api/interpreter'
        req = Request(upstream_url, data=query.encode('utf-8'), headers={'Accept': 'application/json', 'Content-Type': 'text/plain; charset=utf-8', 'User-Agent': 'walker-streets-proxy/1.0'}, method='POST')
        try:
            with urlopen(req, timeout=60) as resp:
                text = resp.read().decode('utf-8')
            data = json.loads(text)
            self._send_json(200, data)
        except Exception as exc:
            self._send_json(500, {'error': 'Unable to fetch street data', 'details': str(exc)})


if __name__ == '__main__':
    server = HTTPServer((HOST, PORT), ProxyHandler)
    print(f'Proxy server listening on http://{HOST}:{PORT}')
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()
